lib/parachain/candidate_validation.go

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package parachain
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	parachaintypes "github.com/ChainSafe/gossamer/lib/parachain/types"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+var (
+	ErrValidationCodeMismatch   = errors.New("validation code hash does not match")
+	ErrValidationInputOverLimit = errors.New("validation input is over the limit")
+)
+
+// PoVRequestor gets proof of validity by issuing network requests to validators of the current backing group.
+// TODO: Implement PoV requestor
+type PoVRequestor interface {
+	RequestPoV(povHash common.Hash) PoV
+}
+
+func getValidationData(runtimeInstance RuntimeInstance, paraID uint32,
+) (*parachaintypes.PersistedValidationData, *parachaintypes.ValidationCode, error) {
+
+	var mergedError error
+
+	for _, assumptionValue := range []scale.VaryingDataTypeValue{
+		parachaintypes.IncludedOccupiedCoreAssumption{},
+		parachaintypes.TimedOutOccupiedCoreAssumption{},
+		parachaintypes.Free{},
+	} {
+		assumption := parachaintypes.NewOccupiedCoreAssumption()
+		err := assumption.Set(assumptionValue)
+		if err != nil {
+			return nil, nil, fmt.Errorf("getting assumption: %w", err)
+		}
+		persistedValidationData, err := runtimeInstance.ParachainHostPersistedValidationData(paraID, assumption)
+		if err != nil {
+			mergedError = errors.Join(mergedError, err)
+			continue
+		}
+
+		validationCode, err := runtimeInstance.ParachainHostValidationCode(paraID, assumption)
+		if err != nil {
+			return nil, nil, fmt.Errorf("getting validation code: %w", err)
+		}
+
+		return persistedValidationData, validationCode, nil
+	}
+
+	return nil, nil, fmt.Errorf("getting persisted validation data: %w", mergedError)
+}
+
+// ValidateFromChainState validates a candidate parachain block with provided parameters using relay-chain
+// state and using the parachain runtime.
+func ValidateFromChainState(runtimeInstance RuntimeInstance, povRequestor PoVRequestor,
+	candidateReceipt parachaintypes.CandidateReceipt) (
+	*parachaintypes.CandidateCommitments, *parachaintypes.PersistedValidationData, bool, error) {
+
+	persistedValidationData, validationCode, err := getValidationData(runtimeInstance, candidateReceipt.Descriptor.ParaID)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("getting validation data: %w", err)
+	}
+
+	// check that the candidate does not exceed any parameters in the persisted validation data
+	pov := povRequestor.RequestPoV(candidateReceipt.Descriptor.PovHash)
+
+	// basic checks
+
+	// check if encoded size of pov is less than max pov size
+	buffer := bytes.NewBuffer(nil)
+	encoder := scale.NewEncoder(buffer)
+	err = encoder.Encode(pov)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("encoding pov: %w", err)
+	}
+	encodedPoVSize := buffer.Len()
+	if encodedPoVSize > int(persistedValidationData.MaxPovSize) {
+		return nil, nil, false, ErrValidationInputOverLimit
```

Comment:
```suggestion
		return nil, nil, false, fmt.Errorf("%w, limit: %d, got: %d", ErrValidationInputOverLimit, persistedValidationData.MaxPovSize, encodedPoVSize)
```

Reply:
I can make this change, but I don't think it serves much purpose to print limit. It will could only get used in very rare scenario during debugging.